### PR TITLE
Remove re-exports of Control.Monad, ...

### LINKED ISSUE #68: MTL doesn't adhere to PVP as it re-exports e.g. Control.Monad
This is subtle issue. But when you do

```
import Prelude ()
import Prelude.Compat --base-compat-0.11

import Control.Monad.Reader -- mtl-2.2.2
```

Then the `fail` is ambiguous.

it's my own fault, that I imported `Control.Monad.Reader` unqualified, but still this is surprising and nasty.

### PR DESCRIPTION
**Pull request by phadej**
... Control.Monad.Fix and Data.Monoid modules.

Import Identity and MonadIO explicitly
in Control.Monad.Identity and Control.Monad.Trans
to avoid potential future API fluctuations.

Resolves #68

### CODE REVIEW

**Comment by phadej** [comment]
I'd actually prefer to not re-export `MonadTrans` either, especially as whether it has `QuantifiedConstraint` depends on the underlying `transformers` version (as non-transformers-0.6 are also allowed). But maybe that can wait for the next major version.

File: Control/Monad/Identity.hs [hunk 1]
```diff
@@ -35,13 +35,7 @@ version of that monad.
 module Control.Monad.Identity (
     module Data.Functor.Identity,
     module Control.Monad.Trans.Identity,
-
-    module Control.Monad,
-    module Control.Monad.Fix,
    ) where
 
-import Data.Functor.Identity
+import Data.Functor.Identity (Identity, runIdentity)
```
Comment (RyanGlScott) [inline Control/Monad/Identity.hs:40]: After this change, `Control.Monad.Identity` no longer re-exports the `Identity` data constructor (only the type constructor). Was this intended? (I ran into this issue in practice when trying to test some packages for `mtl-2.3` compatibility.)

Comment (phadej) [reply Control/Monad/Identity.hs:40]: No. Let me fix this

Comment (phadej) [reply Control/Monad/Identity.hs:40]: That said, I'd recommend to import `Identity` from `Data.Functor.Identity`.

File: Control/Monad/Identity.hs [hunk 2]
```diff
@@ -35,13 +35,10 @@ version of that monad.
 module Control.Monad.Identity (
     module Data.Functor.Identity,
     module Control.Monad.Trans.Identity,
-
-    module Control.Monad,
-    module Control.Monad.Fix,
    ) where
 
-import Data.Functor.Identity
+-- We do this so, because transformers-0.4.0.0
```
Comment (emilypi) [inline Control/Monad/Identity.hs:40]:
```suggestion
-- We do this because transformers-0.4.0.0
```

**Review by emilypi — CHANGES_REQUESTED** [review]

**Comment by emilypi** [comment]
@phadej approved with one small grammatical wibble

**Comment by chessai** [comment]
LGTM pending Emily's comments.

**Comment by sjakobi** [comment]
I've noticed some breakage to this change. See #101.

**Comment by phadej** [comment]
E.g. `lucid` imports `MonadIO` through `Control.Monad.Reader`. This is IMHO surprising.

Having an explicit set list of re-exports is an option too. Though #68 would repeat when e.g. `base-compat` exports `MonadFail.fail` and `Control.Monad.Reader` would export old `Control.Monad.fail` (on `base` where it exists).

File: Control/Monad/Writer/CPS.hs [hunk 3]
```diff
@@ -36,23 +36,16 @@ module Control.Monad.Writer.CPS (
...
-    module Control.Monad,
-    module Control.Monad.Fix,
     module Control.Monad.Trans,
-    module Data.Monoid,
   ) where
 
 import Control.Monad.Writer.Class
 
-import Control.Monad.Trans
+import Control.Monad.Trans (MonadTrans (lift))
```
Comment (phadej) [inline Control/Monad/Writer/CPS.hs:44]: `Control.Monad.Trans` used to re-export `MonadIO(liftIO)` too. I don't think that is a good idea. `MonadTrans(lift)` is somewhat justified, but `MonadIO` seems to be there out-of-blue.

**Comment by phadej** [comment]
Re explicit export list: Note that `Control.Monad` has functions like:

```haskell
msum :: MonadPlus m => [m a] -> m a
```

which changed its type to

```haskell
msum :: (MonadPlus m, Foldable t) => f (m a) -> m a
```

also `foldM` is similarly `Foldable`-generalized. Also `forever`, `guard` are generalized to `Applicative` / `Alternative`.

Also the `<$!>` is present since base-4.8.0.0 only.

We could either:
1. not re-export these functions
2. redefine them to always have "new" types
3. drop support from pre-AMP GHCs.

All of these would still preserve a potential problem that a future `base` change would force a major version of `mtl` due change in type. (Or `mtl` could continue to re-export the name with an old type, if possible).

---

TBH, I looks like there are only bad options
- Try to not break downstream, making maintenance harder (the `base <5` bound won't cut, has to be reviewed every major `base` release, if there are explicit re-export lists)
- Annoy folks buy cleaning up the API surface.

Long term the latter is better, but right now is quite a bad moment for that :(

**Comment by sjakobi** [comment]
I think it would be fine to drop support for some older GHCs, eg 7.x. If
type signatures still shift, we should try to find the most compatible one,
probably by testing with some affected packages.

Oleg Grenrus ***@***.***> schrieb am Fr., 19. Nov. 2021,
16:35:

**Comment by phadej** [comment]
As I said, this will just delay the "inevitable". Folks shouldn't import `Control.Monad.State` etc unqualified and also expect to get `MonadIO` from there. That's unhygienic and *bad* API design.

I'd rather do this right at once. If now is a bad moment (due `Eq` drama and no decision on how much breakage is acceptable, e.g.) I'd rather revert this commit (an re-open #68) than go for suboptimal middle-ground. Still re-exporting something is a baby step which doesn't really bring closer to resolving #68.

**Comment by sjakobi** [comment]
> Still re-exporting something is a baby step which doesn't really bring closer to resolving #68.

I don't understand this. It seems quite viable to find a set of re-exports that would substantially reduce the breakage I reported in #101, and at the same time avoid the PVP problems reported in #68.

In principle I agree that `mtl` shouldn't have these re-exports at all, but the current situation is that lots of packages depend on these now.

I'm not fundamentally opposed to getting rid of the re-exports eventually, but I think it should be done separately from the migration to `transformers-0.6`. Otherwise this migration will be extremely slow and painful.

**Comment by phadej** [comment]
> Otherwise this migration will be extremely slow and painful.

This is "break big but rarely" vs. "break often in small parts" trade-off. In general people prefer no breakages at all.

> Otherwise this migration will be extremely slow and painful.

For the removal of re-exports change the migration is adding 

```haskell
import Control.Monad
```

like import on top of

```haskell
import Control.Monad.Reader
```

Not painful. Requires more packages to be patched, but if we are going to do that anyway, why not now?

Anyway. I won't write a patch for a *set of re-exports that...*, and I'm fine with this patch being reverted if it's a blocker for release of `mtl`. But this is up to actual maintainers of `mtl` whom have been silent for now. I think we discussed this enough so they can make a call what to do.

**Pull request merged**